### ISSUE TITLE
Token preemptively completing experiments...

### ISSUE THREAD

**Issue by danieljwilson** [expfactory/expfactory#160]
## Version of Experiment Factory:

Latest

### Expected behavior

Following the guide for [Usage > Use Tokens](https://expfactory.github.io/usage#start-a-headless-experiment-container) I attempted to provide a pre-generated URL using the suggested format:

```html
https://<your-server>/login?token=<token>
```

I expected to start the battery of experiments with the results saved to the folder associated with the provided token.

### Actual behavior

What happened instead was that a banner appeared saying that I had completed all the experimnets:

<img width="951" alt="image" src="https://user-images.githubusercontent.com/7377893/159235973-a64572f0-a27b-4882-9ef7-609ac7ff20ff.png">

Also the user folder on the server was also marked as `_finished` (appended to the folder token name)

### Steps to reproduce behavior

You can run from my [public repo](https://github.com/danieljwilson/thesis_battery_sona) - also I have tried this with various combinations of experiments and the behavior has occurred each time.

**Comment by vsoch** [comment]
So to clarify this only happens with adding the token to the URL, and not when you provide directly in the interface?

**Comment by danieljwilson** [comment]
Yes this happens when I add the token to the URL. Sorry not sure what that means to provide it directly in the interface? Do you mean in interactive mode?

**Comment by vsoch** [comment]
yep! So normally there is a box you can enter it in (not in the browser url), and I would call that some kind of interactive mode. Does everything work okay with that method?

**Comment by danieljwilson** [comment]
I will try that and report back!

**Comment by vsoch** [comment]
I'm in the middle of the work day so I can't debug for you, but I can tell you what I think is going on (and how to debug!) So this function: https://github.com/expfactory/expfactory/blob/1b0b570f64355fe32ddca071bd0af5eb267e4aa2/expfactory/views/main.py#L119-L143

It looks like the last block is being skipped over (no experiment found) and then it's going directly to finish. So what we are going to do is put a ton of print statements around there, and install from your local version to build the container. E.g., from your repo

```
git clone https://github.com/expfactory/expfactory
```
And then in your Dockerfile instead of:

```
WORKDIR /opt 
RUN git clone -b master https://github.com/expfactory/expfactory
WORKDIR expfactory 
```
Do:
```
WORKDIR /opt 
COPY ./expfactory /opt/expfactory
WORKDIR expfactory 
```
And then in that function, try something like this:

```python
@app.route("/next", methods=["POST", "GET"])
def next():

    # Headless mode requires logged in user with token
    if app.headless and "token" not in session:
        return headless_denied()

    # To generate redirect to experiment
    experiment = app.get_next(session)

    # I suspect you are going to see None here!
    print(experiment)
    print(experiment is not None)
    if experiment is not None:
        app.logger.debug("Next experiment is %s" % experiment)
        template = "/experiments/%s" % experiment

        # Do we have runtime variables?
        token = session.get("token")
        if app.vars is not None:
            variables = get_runtime_vars(
                token=token, varset=app.vars, experiment=experiment
            )
            template = "%s?%s" % (template, variables)

        print("We got to performing checks, next experiment is %s" % experiment)
        return perform_checks(template=template, do_redirect=True, next=experiment)

    # But I think you are skipping the above and hitting down here
    return redirect("/finish")
```
so then the question is - why is app.get_next() returning there aren't experiments? That bit is here: https://github.com/expfactory/expfactory/blob/1b0b570f64355fe32ddca071bd0af5eb267e4aa2/expfactory/server.py#L99-L114. I suspect it's because there aren't experiments in the session, so this will be empty: https://github.com/expfactory/expfactory/blob/1b0b570f64355fe32ddca071bd0af5eb267e4aa2/expfactory/server.py#L107. For that, I would then check this function: https://github.com/expfactory/expfactory/blob/1b0b570f64355fe32ddca071bd0af5eb267e4aa2/expfactory/server.py#L71-L97. Note that since we pipe the log into a file (and then /dev/null to keep the container running) you will likely want to tweak startscript.sh so it just prints to the console and you can use `docker logs` to see it. And I've been wanting to make it easier to see logs in that respect, so if you find a change that works please open a PR! Keep me updated!

**Comment by danieljwilson** [comment]
This is pretty great "middle-of-workday" feedback! 

I am in France right now so heading to bed atm but will take a look at this in the morning!

**Comment by vsoch** [comment]
Any progress or updates here?

**Comment by danieljwilson** [comment]
I haven't tried this is a while - have been giving people tokens for them to enter instead. I have this semi-automated using airtable automations (in terms of emailing each participant their token) so this has been my workaround.

**Comment by vsoch** [comment]
Safe to close then until someone reproduces?

**Comment by danieljwilson** [comment]
I think that is fine! Thanks for the follow up :)

**Comment by AlvaroAguilera** [comment]
I can reproduce this issue in the current version.

**Comment by vsoch** [comment]
oh that's good! Can you use my previous instructions https://github.com/expfactory/expfactory/issues/160#issuecomment-1074358564 to debug? Re-opening too.
